internal/service/resources/collector/collector.go

```diff
@@ -0,0 +1,229 @@
+			Name:         name,
+			GenerationID: dataSource.GetGenerationID(),
+		})
+		if err != nil {
+			return fmt.Errorf("failed to create new data source %q: %w", name, err)
+		}
+
+		dataSource.SetID(*result.DataSourceID)
+		slog.Info("created new data source", "name", name, "uuid", *result.DataSourceID)
+	case err != nil:
+		return fmt.Errorf("failed to get data source %q: %w", name, err)
+	default:
+		dataSource.SetID(*record.DataSourceID)
+		dataSource.SetGenerationID(record.GenerationID)
+		slog.Info("restored data source",
+			"name", name, "uuid", record.DataSourceID, "generation", record.GenerationID)
+	}
+	return nil
+}
+
+// execute runs a single iteration of the main loop.  It does not return an error because all errors should be handled
+// gracefully.  If a truly unrecoverable error happens then a panic should be used to restart the process.
+func (c *Collector) execute(ctx context.Context) {
+	slog.Debug("collector loop running", "sources", len(c.dataSources))
+	for _, d := range c.dataSources {
+		slog.Debug("collecting data from data source", "source", d)
+		if err := c.executeOneDataSource(ctx, d); err != nil {
+			slog.Warn("failed to collect data from data source", "source", d, "error", err)
+		} else {
+			slog.Debug("collected data from data source", "source", d)
+		}
+	}
+	slog.Debug("collector loop complete", "sources", len(c.dataSources))
+}
+
+// executeOneDataSource runs a single iteration of the main loop for a specific data source instance.
+func (c *Collector) executeOneDataSource(ctx context.Context, dataSource DataSource) (err error) {
+	// Increment the generation id
+	dataSource.IncrGenerationID()
+
+	// Get the list of resource pools for this data source
+	pools, err := c.collectResourcePools(ctx, dataSource)
+	if err != nil {
+		return fmt.Errorf("failed to collect resource pools: %w", err)
+	}
+
+	// Get the list of resources for this data source
+	_, err = c.collectResources(ctx, dataSource, pools)
+	if err != nil {
+		return fmt.Errorf("failed to collect resources: %w", err)
+	}
+
+	// TODO: purge stale record
+
+	// TODO: persist data source info
+
+	return nil
+}
+
+// collectResources collects Resource objects from the data source, persists them to the database,
+// and signals any change events to the notification processor.
+func (c *Collector) collectResources(ctx context.Context, dataSource DataSource,
+	pools []models.ResourcePool) ([]models.Resource, error) {
+	resources, err := dataSource.GetResources(ctx, pools)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get resources: %w", err)
+	}
+
+	// Loop over the set of resources and create the associated resource types.
+	seen := make(map[uuid.UUID]bool)
+	for _, resource := range resources {
+		resourceType, err := dataSource.MakeResourceType(&resource)
+		if err != nil {
+			return nil, fmt.Errorf("failed to make resource type from '%v': %w", resource, err)
+		}
+
+		if _, ok := seen[resourceType.ResourceTypeID]; ok {
```

Comment:
if seen[resourceType.ResourceTypeID] is enough. Not need to use ok

Reply:
Fixed